libs/core/algorithms/include/hpx/parallel/util/loop.hpp

```diff
@@ -38,23 +39,66 @@ namespace hpx::parallel::util {
             HPX_HOST_DEVICE HPX_FORCEINLINE static constexpr Begin call(
                 Begin it, End end, F&& f)
             {
+                return call(hpx::execution::seq, it, end, HPX_FORWARD(F, f));
+            }
+
+            template <typename Begin, typename End, typename CancelToken,
+                typename F,
+                HPX_CONCEPT_REQUIRES_(hpx::traits::is_iterator_v<End>&&
+                        hpx::traits::is_forward_iterator_v<Begin>)>
+            HPX_HOST_DEVICE HPX_FORCEINLINE static Begin call(
+                Begin it, End end, CancelToken& tok, F&& f)
+            {
+                return call(
+                    hpx::execution::seq, it, end, tok, HPX_FORWARD(F, f));
+            }
+
+            template <typename ExPolicy, typename Begin, typename End,
```

Comment:
Should we mix and match vectorized and non-vectorized implementations in the same file? For other algorithms we have split those?

Reply:
I personally believe we should try to have both unseq and seq loops in the same file to allow for easier dispatch.

`template <typename ..... , HPX_CONCEPT_REQUIRE(vectorization_condition)>
//vector loop

template <typename ..... , HPX_CONCEPT_REQUIRE(!vectorization_condition)>
//non-vectorised loop
`
seems like a good way to do things according to me and having both in the same file might be simply easier to work with in the future instead of jumping around files.

Would like to hear your opinion too @hkaiser, do you think we should rather have 2 different files and depracate `#include <hpx/parallel/util/loop.hpp>` and `#include <hpx/parallel/unseq/loop.hpp>` in favour of single header which implicitly includes both the files?

Reply:
Having it in one or two files doesn't make a difference indeed. What I really meant was that the way you dispatched between the vectorized and non-vectorized loops involves a two level dispatch mechanism. I'd rather have one dispatch point (which is also used for the datapar algorithm implementations). Currently (without unseq) we dispatch between non-vectorized and datapar implementations using e.g. the `loop_t` customization point. 

Instead of adding another dispatch point behind this using a different mechanism (you're suggesting to SFINAE), I think it would be better to provide `tag_invoke` overloads for e.g. the `loop_t` CPO as done for datapar - for an example see: https://github.com/STEllAR-GROUP/hpx/blob/09eabc71da8f2a150b2e26efe3d52d4163e49072/libs/core/algorithms/include/hpx/parallel/datapar/loop.hpp#L417-L422

Reply:
I'm not saying you should introduce a new customization point. What I meant is that you should use the already existing CPOs by providing appropriate `tag_invoke` overloads that dispatch to your code.

Reply:
I am sorry but I don't seem to understand the issue completely yet

https://github.com/STEllAR-GROUP/hpx/blob/15e89648b5f41002ea4ff41a9b4b42e9c2b700e5/libs/core/algorithms/include/hpx/parallel/util/loop.hpp#L37-L54

Is the concern on why we are calling `call(hpx::execution::seq, ...)` from `call(...)`. If that is the case the reason I kept it is such as to not raise errors is cases where `detail::loop<Iter>().call(...)`. If required I can remove the overload and try to push it to tag_fallback_invoke over here
https://github.com/STEllAR-GROUP/hpx/blob/15e89648b5f41002ea4ff41a9b4b42e9c2b700e5/libs/core/algorithms/include/hpx/parallel/util/loop.hpp#L106-L127

Reply:
My concern is that you have two sets of overloads for the `detail::loop::call` functions. One for the vectorized and one for the non-vectorized implementation. That means that the execution has to go through two customization points: a) the `loop_t` selecting between `simd` and non-`simd` and b) the mentioned two sets of `call` overloads. I suggest to dispatch only through the `loop_t` customization point object.